### ISSUE TITLE
iscsi digests tests broken on nightly.

### ISSUE THREAD

**Issue by Seth5141** [spdk/spdk#963]
## Expected Behavior
Test passes

## Current Behavior
The last several nightly tests have failed on both the rbd and iscsi jobs in the digests tests. See logs here:
https://dqtibwqq6s6ux.cloudfront.net/results/autotest-nightly/builds/520/archive/RBD_autotest_nightly/index.html
https://dqtibwqq6s6ux.cloudfront.net/results/autotest-nightly/builds/520/archive/iscsi_autotest_nightly/index.html
https://dqtibwqq6s6ux.cloudfront.net/results/autotest-nightly/builds/519/archive/RBD_autotest_nightly/index.html
https://dqtibwqq6s6ux.cloudfront.net/results/autotest-nightly/builds/519/archive/iscsi_autotest_nightly/index.html

## Possible Solution
I believe the regression was caused by this patch: https://github.com/spdk/spdk/commit/80cf038cedc0399b01c352b9dc9a1047836117ce
and I have a patch out here:
https://review.gerrithub.io/c/spdk/spdk/+/468778
to revert it.
After reverting the above patch, I was able to successfully run the RBD and iscsi tests.
https://dqtibwqq6s6ux.cloudfront.net/public_build/autotest-nightly_521.html (link may not be active yet as the test results are still copying to the server at the time of writing this post).

## Steps to Reproduce
1. Run the iscsi digests tests on master.

## Context (Environment including OS version, SPDK version, etc.)
Fails on all of the relevant CI machines.

**Comment by jimharris** [comment]
I think just reverting this patch is best for now.  It's not clear why the patch breaks the digest test yet.  But we also don't know for sure yet that this patch is needed for iSCSI optimization.  It may just be faster and simpler to just copy the 48-byte BHS.  I think we should wait to merge an updated version of the reverted patch until we have clear proof that this patch is needed.

**Comment by changpe1** [comment]
There is one more place need to be added if changed the bhs field in PDU to a pointer:

diff --git a/lib/iscsi/iscsi.c b/lib/iscsi/iscsi.c
index 05ac33b..28e0660 100644
--- a/lib/iscsi/iscsi.c
+++ b/lib/iscsi/iscsi.c
@@ -286,7 +286,7 @@ spdk_iscsi_pdu_calc_header_digest(struct spdk_iscsi_pdu *pdu)
        uint32_t ahs_len_bytes = pdu->bhs->total_ahs_len * 4;

        crc32c = SPDK_CRC32C_INITIAL;
-       crc32c = spdk_crc32c_update(&pdu->bhs, ISCSI_BHS_LEN, crc32c);
+       crc32c = spdk_crc32c_update(pdu->bhs, ISCSI_BHS_LEN, crc32c);

        if (ahs_len_bytes) {
                crc32c = spdk_crc32c_update(pdu->ahs, ahs_len_bytes, crc32c);

I changed one line, and it works for Ziye's patch, anyway reverting this patch is fine to me.

**Comment by jimharris** [comment]
Good catch Changpeng.  I still suggest we just revert this patch for now.

**Comment by optimistyzy** [comment]
Hi Changpeng,

Thanks for catching the root cause of this issue, it should be the right
fix.

在 2019年9月19日星期四，Jim Harris <notifications@github.com> 写道：

**Comment by benlwalker** [comment]
I'd also like to confirm that avoiding this copy actually improves performance. Similarly in the NVMe-oF/TCP code, I'd like to confirm that it actually helps. I'm interested for two reasons:

1) This is making the code more complicated. It would be nice to simplify.
2) If we can tolerate this small copy, we could actually move the 8k read buffering into the socket abstraction and only code that up once and get it automatically in all places we use the socket layer.

**Comment by jimharris** [comment]
Bug scrub 09/24: Closing this issue since iscsi digest tests now working.  We will determine separately whether to merge an updated version of this patch, after we know if it helps with performance as we optimize kernel TCP usage with SPDK.
